### ISSUE TITLE
Release KTX 1.10.0-b3

### ISSUE THREAD

**Issue by czyzby** [libktx/ktx#381]
- [x] Maven Central upload.
- [x] GitHub release.

---

This release includes a new module with reflection utilities: `ktx-reflect`. It also has some notable new features in `ktx-collections`, `ktx-inject`, and `ktx-assets-async`, with the new `AsyncAssetManager` being a major one.

**Comment by czyzby** [comment]
@vrudas, @runt9, @Quillraven: Do you plan on submitting any new pull requests? I can hold on with the release a bit.

As a side note, I'd appreciate if you gave [`AsyncAssetManager`](https://github.com/libktx/ktx/blob/develop/assets-async/src/main/kotlin/ktx/assets/async/manager.kt#L40) a try and left your feedback. `AssetStorage` is still the advised class for asset management, but I think it's worth it to have a drop-in replacement for `AssetManager` with some extra features. Here's the [documentation](https://github.com/libktx/ktx/tree/develop/assets-async#asyncassetmanager-guide).

**Comment by Quillraven** [comment]
No PRs from my side coming and sorry, but currently I don't have time to work on a game or update an existing one. That's why I will not be able to give feedback to AsyncAssetManager soon. But I will keep that in mind for a future project!

Not sure if it is for interest of you but I saw the "help wanted" for artemis extensions a few weeks ago. That motivated me again to work on my Kotlin ECS (=KECS repository). But I started again from scratch and learned a lot the last times when I had time to work on it.

I will try to publish a workable draft by end of this week and will add you as reviewer. Would be great if you have time to give some feedback since you are smarted than me especially when it comes to Kotlin ;)

A guy from my stream mentioned that the Korge developer might be interested in that as well and it could be a "native Kotlin" alternative for LibKTX as well?

**Comment by runt9** [comment]
@czyzby No PRs incoming from me at the moment. I'm still in the early stages of rewriting my game from the ground up with a more stable architecture. That said, I think the "binding" architecture might end up being it's own thing and _possibly_ a good candidate for a KTX module, though it's a fairly opinionated "framework" kinda thing as opposed to a library.

I just got finished implementing the AssetStorage solution, so I'm not too far removed from it and could probably easily go back and take a look at AsyncAssetManager. That'd probably be tomorrow as I'm working on getting my settings bindings and MVC+S flow working for that right now. I'm certainly happy to give it a look, though. I'll let you know my thoughts sometime tomorrow.

If you're curious about the "binding" piece of what I'm working on, here's the class from the current working branch (more changes locally from today I haven't pushed like an auto-evaluating dirty flag that allows things like auto disabling/enabling a button based on dirty flag, like an apply button in a settings dialog): https://github.com/firstrateconcepts/FusionOfSouls/blob/working/src/main/kotlin/net/firstrateconcepts/fusionofsouls/util/framework/ui/ViewModel.kt

**Comment by czyzby** [comment]
@Quillraven 

> A guy from my stream mentioned that the Korge developer might be interested in that as well and it could be a "native Kotlin" alternative for LibKTX as well?

Yeah, Korge is a fun project. I'm a bit biased as a KTX developer, but I think there's place for both, given than LibGDX is more stable and popular than Korge in general, so it benefits a lot from Kotlin bindings.

That being said, I don't see why a native Kotlin ECS library wouldn't be useful for both, as long as you make sure that it works as a multiplatform library and don't rely on any JVM APIs in the core module to conform with Korge.

> How is the `getProgress` working for loading boards? Is it the same as with AssetManager? I guess not because of the asynchronous loading? Is it then using the same/similar approach as AssetStorage?

Exactly the same as with the basic `AssetManager`. `AssetManager.load` also provides asynchronous loading in the sense that it doesn't freeze your app and does a lot of operations on a different thread. The main different between `load` and `loadAsync` in `AsyncAssetManager` is that `load` returns nothing and requires you to extract the loaded asset from the manager, while `loadAsync` returns `Deferred<Asset>` that allows you to `await` on a particular asset without manually having to call `get`. `Deferred` will also rethrow any loading errors, which improves error handling by a lot.

The initial problems with `AssetStorage` progress reporting were related to synchronizing coroutines with the rendering code - you could potentially start asking the service for progress before the assets were fully scheduled for loading by a coroutine, since these things could happen in parallel. `AssetStorage.loadAsync` being callable outside of coroutines helps with that, and the current progress reporting is even better than in the case of the `AssetManager`, since it also counts dependencies of the scheduled assets. To make sure that the progress is accurate, you just have to schedule loading of your assets with `loadAsync` in `create` or equivalent initialization method. Also, adding some animation/smoothing to the progress bar for a good measure doesn't hurt. :)

> I will try to publish a workable draft by end of this week and will add you as reviewer. Would be great if you have time to give some feedback since you are smarted than me especially when it comes to Kotlin ;)

Don't say it like that, your KTX tutorials are awesome! :)

---

@runt9

> I just got finished implementing the AssetStorage solution, so I'm not too far removed from it and could probably easily go back and take a look at AsyncAssetManager.

Glad to hear that. Just so you know, `AssetStorage` is still advised over `AsyncAssetManager`, since it integrates with coroutines better, and is more performant overall. I think of `AsyncAssetManager` as a drop-in replacement for `AssetManager` that might help some people integrate coroutines into their apps, and eventually migrate to `AssetStorage`.

> That said, I think the "binding" architecture might end up being it's own thing and _possibly_ a good candidate for a KTX module, though it's a fairly opinionated "framework" kinda thing as opposed to a library.

Looking forward to that. If it's a separate thing rather than a replacement of `ktx-inject`, or if it will use `ktx-inject` internally, we could consider adding it as an extra module, but I'm a bit afraid of maintaining native classpath scanning for each backend, since I know it was a bit problematic the last time I've done something like this.

Not sure if you've already seen the Autumn library in `gdx-lml` repo that I've posted earlier, but this might help you implement stuff like annotation extraction or classpath scanning for each platform.

**Comment by vrudas** [comment]
Hi @czyzby, not plan to submit new PRs in the nearest 2 weeks

**Comment by czyzby** [comment]
Thanks for the heads up guys, I'll release the next version soon.

**Comment by czyzby** [comment]
CI failed on `master` due to a missing Dokka dependency after removal of the `jcenter` repository. I'll update Dokka version in a fix branch to make sure that the documentation is published.

**Comment by runt9** [comment]
@czyzby Ok good to know that AssetStorage is preferred, I'll stick with that then. Also for the "binding" library concept, it's actually unrelated to the dependency injection question, it actually has to do with an MVVM concept of being able to have Views auto-update when changes occur in their associated ViewModels, and the changes are initiated by the Controller of the ViewModel.

An example in my game will be the Player's Gold total. There's a bar at the top of the screen with a label for Gold: <amount>. The concept allows the view to bind to the gold amount in the ViewModel and the Controller receives a `GoldUpdatedEvent` or something along those lines from the EventBus which updates the gold variable in the VM which propagates immediately to the VisLabel via some binding functions under-the-hood. It's been a lot more seamless than I initially thought and really helps me enforce a separation of concerns between the different components of Controller, View, ViewModel, Service, and Model.

**Comment by czyzby** [comment]
Oh yeah, that's an interesting feature. I don't think libGDX GUI supports anything like that out of the box, so it would be a nice addition.
